src/mixin/TransientStorageArray.sol

```diff
@@ -0,0 +1,111 @@
+/// in the transient storage slot immediately before. The endianness of
+/// each word has not been changed.
+/// The array is defined from the end to avoid conflicts with transient
+/// variables that could be relied on by the importing contract. If the contract
+/// that inherits this one only access transient storage through Solidity
+/// variables that are only manipulated through basic Solidity operations (i.e.,
+/// no assembly) then there should be no interferences with this contract.
+/// Otherwise, it's important to have a full understanding of the transient
+/// storage layout of the inheriting contract to avoid that the same transient
+/// slot is overridden when working with two unrelated variables.
+abstract contract TransientStorageArray {
+    uint256 constant BYTES_IN_WORD = 32;
+
+    /// @notice The number of bytes currently stored in the transient storage
+    /// array.
+    uint256 internal transient transientStorageLength;
+
+    function storeToTransientStorageArray(bytes calldata data) internal {
+        transientStorageLength = data.length;
+        uint256 storedBytes = 0;
+        while (storedBytes < data.length) {
+            // Each (transient) storage slot stores a full word.
+            bytes32 window;
+            // Note: if windowEnd > data.length, then the extra bytes of
+            // `window` are filled with arbitrary bytes (or zero bytes if after
+            // the calldata range). The variable `length` encodes the array
+            // boundaries, meaning that these extra bytes will be ignored when
+            // recovering the stored content.
+            assembly ("memory-safe") {
+                // `add` is not expected to overflow as sending a transaction
+                // with enough calldata to cause the offset to overflow would
+                // not fit the block gas limit.
+                window := calldataload(add(data.offset, storedBytes))
+            }
+
+            uint256 transientArraySlot;
+            unchecked {
+                // No overflow check: block gas limit avoids overflows.
+                transientArraySlot = type(uint256).max - (storedBytes / BYTES_IN_WORD);
+            }
+
+            assembly ("memory-safe") {
+                tstore(transientArraySlot, window)
+            }
+            unchecked {
+                // No overflow check: block gas limit avoids overflows.
+                storedBytes += BYTES_IN_WORD;
+            }
+        }
+    }
+
+    function readTransientStorageArray() internal view returns (bytes memory data) {
+        uint256 byteLength = transientStorageLength;
+        // The number of words to read is `ceil(byteLength / BYTES_IN_WORD)`.
+        // `byteLength` could be controlled by the user, so we check for
+        // addition overflows. All other operations are known not to cause
+        // over/underflows.
+        uint256 wordsToRead = byteLength + BYTES_IN_WORD;
```

Comment:
Doesn't the comment in the store function say it's impossible for byteLength to be so big that it causes overflows because that would exceed the block's gas limit? Why doesn't that apply here, too?